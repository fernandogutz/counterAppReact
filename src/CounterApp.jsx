/* 
----------------------
Tarea
-----------------------
1. Crear un nuevo componente dentro de la carpeta SRC llamado CounterApp

2. El CounterApp debe de ser un Functional Component

3. El contenido del CounterApp debe de ser:

    <h1>CounterApp</h1>
    <h2> { value } </h2>

4. Donde "value" es una propiedad enviada desde el padre hacia el componente CounterApp (Debe ser númerica validada con PropTypes)

5. Reemplazar en el index.js ó main.jsx el componente de <PrimeraApp /> por el componente <CounterApp /> (no se olviden del value que debe de ser un número)

6. Asegúrense de no tener errores ni warnings (Cualquier warning no usado, comentar el código)
*/
import { useState } from 'react'; //Importando nuestro primer hook
import PropTypes from 'prop-types';

const handleAddTest = (event, name) => { // Al declarar la función fuera del componente, no se asignará un espacio en memoria cada vez que se renderice el elemento
    console.log(name);  
    console.log(event);  
    //console.log(value);  // Esto da error, si queremos usarlo debemos poner la función dentro del functional component
} 


const CounterApp = ( {value} ) => {

    console.log('render'); //Cada vez que hay un cambio en el estado, se dispara el functional component, es decir, se renderiza todo el componente nuevamente

    const [ counter, setCounter ] = useState( value );
    
    const handleAdd = () => {
        setCounter( counter + 1 );
        //setCounter( (c) => c + 1 ); // Si no tenemos referencia a counter podríamos hacer la suma así con una función
    }

    const handleSubstract = () => {
        setCounter(counter - 1);
    }
    
    const handleReset = () => {
        setCounter(value);
    }

    return (
        <>
            <h1>CounterApp</h1>
            <h2> { counter } </h2>  
            <button onClick={ (event) => handleAddTest(event, 'Fer', counter) }>
                Test Evento click
            </button>

            <button onClick={ handleAdd }> +1 </button>         
            <button onClick={ handleSubstract }> -1 </button>         
            <button onClick={ handleReset }> Reset </button>         
        </>
    );
}

CounterApp.propTypes = {
    value: PropTypes.number
}

CounterApp.defaultProps = {
    value: 0
}

export {
    CounterApp
}

