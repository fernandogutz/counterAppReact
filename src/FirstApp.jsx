import PropTypes from 'prop-types';

const newMsg = {
    msg: 'Hola',
    name: 'fer'
}; // Podemos usar esta variable en el componente
// Entre llaves podemos imprimir datos de js en el componente, excepto objetos, para ello usamos JSON.stringify(objetoJs)

// Tarea: Retornar una función en el componente
const saludar = (name) => `Hola ${name}`; // Lo dejamos fuera del componente para que no se use este espacio en memoria cada vez que se renderice

const FirstApp = ({title, subTitle='sasdasd'}) => {

    return (
        <>
            <h1>{title}</h1>
            <code>{ JSON.stringify(newMsg) }</code>
            <h2>{subTitle}</h2>
        </>
    )
}

export {
    FirstApp
}

// Importamos el paquete con npm install prop-types y lo usamos con PropTypes
FirstApp.propTypes = {
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string.isRequired,
};

//Tomamos nuestro functional component FirstApp y le asignamos valores por defecto a sus proptypes
FirstApp.defaultProps = {
    title: 'No hay title',
    //subTitle: 'No hay subtitle',
    name: 'Fernando Gutz' //También podemos añadir nuevos proptypes
};