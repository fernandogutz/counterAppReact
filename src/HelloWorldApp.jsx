// Usualmente todo en react comienza con un componente (el punto de entrada de nuestra app)
// se acostumbra trabajar hoy en día con componentes funcionales(functional components, en vez de clases)
const App = () => <h1>Hola Mundo</h1>; 

// Los functional components deben ir en UpperCamelCase
// El componente App es el que llamaremos luego con <App/>
// Normalmente el componente inicial no está en el main.js

export {
    App
}