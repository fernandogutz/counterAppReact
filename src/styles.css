html, body {
    background-color: #101010;
    color: #ccc;
    font-family: Verdana, Geneva, Tahoma, sans-serif;
    font-size: 1.3rem;
    padding: 70px;
}

button {
    padding: 15px;
    font-size: 1.3rem;
    margin: 12px;
    margin-left: 0;
}