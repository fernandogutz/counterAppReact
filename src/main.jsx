import React from 'react'; // Primero importamos react
import ReactDOM from 'react-dom/client'; // Luego importamos alguna herramienta para renderizar(dom, electron, native)

import {FirstApp} from './FirstApp';
import {CounterApp} from './CounterApp';
import './styles.css'; // Cargando los estilos globales en el main.jsx

ReactDOM.createRoot( document.getElementById('root')).render(
    <React.StrictMode>
        <CounterApp value={999}/>
    </React.StrictMode>
);
    