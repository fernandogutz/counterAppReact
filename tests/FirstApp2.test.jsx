import { render, screen } from "@testing-library/react"
import { FirstApp } from "../src/FirstApp"

describe('Pruebas en <FirstApp></FirstApp>', () => { 
   const subtitle = 'Hola, soy un subtituloo';

   test('debe hacer macth con el snapshot', () => { 
      const {container} = render(<FirstApp subTitle={subtitle}/> )
      expect(container).toMatchSnapshot();
   });

   test('debe mostrar el mensaje Hola, soy un subtituloo', () => { 
      render(<FirstApp subTitle={subtitle}/> );
      expect( screen.getByText(subtitle) ).toBeTruthy();
   
   });

   test('debe mostar el subtitulo en un h2', () => { 
      render(<FirstApp subTitle={subtitle}/> );
      expect( screen.getByRole('heading', { level: 2 }).innerHTML ).toContain(subtitle);
   })

})