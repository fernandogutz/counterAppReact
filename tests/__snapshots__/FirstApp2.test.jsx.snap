// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <FirstApp></FirstApp> debe hacer macth con el snapshot 1`] = `
<div>
  <h1>
    No hay title
  </h1>
  <code>
    {"msg":"Hola","name":"fer"}
  </code>
  <h2>
    Hola, soy un subtituloo
  </h2>
</div>
`;
